src/swap/defi/thorchain.js

```diff
@@ -0,0 +1,611 @@
+          clog
+        )
+      } else {
+        throw new Error('TODO: TO quoting')
+      }
+      const { fromNativeAmount, toNativeAmount, limit } = calcResponse
+
+      let customNetworkFee
+      let customNetworkFeeKey
+
+      const customFeeTemplate = (fromWallet.currencyInfo.customFeeTemplate ??
+        [])[0]
+      const fromCurrencyInfo = fromWallet.currencyInfo
+      if (customFeeTemplate?.type === 'nativeAmount') {
+        customNetworkFee = gasRate
+        customNetworkFeeKey = customFeeTemplate.key
+      } else if (fromCurrencyInfo.defaultSettings?.customFeeSettings != null) {
+        // Only know about the key 'gasPrice'
+        const usesGasPrice = fromCurrencyInfo.defaultSettings.customFeeSettings.find(
+          f => f === 'gasPrice'
+        )
+        if (usesGasPrice != null) {
+          customNetworkFee = gasRate
+          customNetworkFeeKey = 'gasPrice'
+        }
+      }
+
+      if (customNetworkFee == null || customNetworkFeeKey == null) {
+        throw new SwapCurrencyError(swapInfo, fromCurrencyCode, toCurrencyCode)
+      }
+
+      const memo = buildSwapMemo({
+        chain: toMainnetCode,
+        asset: toCurrencyCode,
+        address: toAddress,
+        limit,
+        affiliateAddress: thorname,
+        points: affiliateFeeBasis
+      })
+
+      const publicAddress = thorAddress
+
+      const spendInfo: EdgeSpendInfo = {
+        currencyCode: request.fromCurrencyCode,
+        spendTargets: [
+          {
+            memo,
+            nativeAmount: fromNativeAmount,
+            publicAddress
+          }
+        ],
+        networkFeeOption: 'custom',
+        customNetworkFee: {
+          [customNetworkFeeKey]: customNetworkFee
+        },
+
+        swapData: {
+          isEstimate: false,
+          payoutAddress: toAddress,
+          payoutCurrencyCode: toCurrencyCode,
+          payoutNativeAmount: toNativeAmount,
+          payoutWalletId: toWallet.id,
+          plugin: { ...swapInfo }
+        },
+        otherParams: {
+          outputSort: 'targets'
+        }
+      }
+
+      const tx: EdgeTransaction = await request.fromWallet.makeSpend(spendInfo)
+
+      return makeSwapPluginQuote(
+        request,
+        fromNativeAmount,
+        toNativeAmount,
+        tx,
+        toAddress,
+        pluginId,
+        false,
+        new Date(Date.now() + EXPIRATION_MS),
+        tx.txid
+      )
+    }
+  }
+  return out
+}
+
+type BuildSwapMemoParams = {
+  chain: string,
+  asset: string,
+  address: string,
+  limit: string,
+  affiliateAddress: string,
+  points: string
+}
+
+const calcSwapFrom = async (
+  params: {
+    fromWallet: EdgeCurrencyWallet,
+    fromCurrencyCode: string,
+    toWallet: EdgeCurrencyWallet,
+    toCurrencyCode: string,
+    nativeAmount: string,
+    sourcePool: Pool,
+    destPool: Pool,
+    volatilitySpreadFinal: string,
+    affiliateFee: string,
+    feeInDestCurrency: string,
+    dontCheckLimits?: boolean
+  },
+  clog: Function
```

Comment:
`clog: EdgeLogMethod`. The `Function` type is pretty much an `any`.

Reply:
Ok. fair enough